### PR TITLE
PUD-106 Added 'get recall document' endpoint

### PR DESCRIPTION
**Pull request by mr-ken**

### CODE REVIEW

**Review by AndyNewton — COMMENTED** [review]
Please see few QQs left.  Thanks.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/controller/RecallsController.kt [hunk 1]
```diff
@@ -78,6 +79,29 @@ class RecallsController(
       .created(URI.create("$baseUri/recalls/$recallId/documents/$fileS3Key"))
       .body(AddDocumentResponse(id = fileS3Key))
   }
+
+  @GetMapping("/recalls/{recallId}/documents/{documentId}")
```
[the remark this answers is not in the document]
Comment (mr-ken) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/controller/RecallsController.kt:83]: We're following the [HMPPS standard for API development and management](https://dsdmoj.atlassian.net/wiki/spaces/NDSS/pages/3159916691/HMPPS+standards+for+API+Development+and+Management) which encourages the use of sub-resources. 
You're right that the document ID is unique but it doesn't have to be. Also it makes it simpler to query for documents ; don't need to add another repository to fetch documents.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/service/RecallDocumentService.kt [hunk 2]
```diff
@@ -16,20 +16,37 @@ class RecallDocumentService(
...
+  }
+
+  private fun recallById(recallId: RecallId) = try {
+    recallRepository.getByRecallId(recallId)
+  } catch (e: EntityNotFoundException) {
+    throw RecallNotFoundError("Recall not found: '$recallId'", e)
+  }
 }
+
+class RecallNotFoundError(message: String, e: Throwable) : Throwable(message, e)
```
[the remark this answers is not in the document]
Comment (mr-ken) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/service/RecallDocumentService.kt:50]: All exceptions in Kotlin are unchecked (no need for RuntimeExceptions) ; I think that one-liner is worth it to make the exception as specific as possible.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/service/RecallDocumentService.kt [hunk 3]
```diff
@@ -16,20 +16,37 @@ class RecallDocumentService(
...
     val fileS3key = s3Service.uploadFile(documentBytes)
     val document = RecallDocument(
       id = fileS3key,
       recallId = recallId.value,
       category = documentCategory
     )
-    recallRepository.save(recall.copy(documents = recall.documents.plus(document)))
+    recallRepository.save(recall.copy(documents = recallById(recallId).documents.plus(document)))
+    // TODO: [KF] delete the document from S3 if saving fails?
     return fileS3key
```
[the remark this answers is not in the document]
Comment (mr-ken) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/service/RecallDocumentService.kt:33]: a) the database table doesn't need to reflect the API (it's completely internal to this app)
b) the column `id` in the table `recall_document` is very clearly the document ID

File: src/test/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/service/RecallDocumentServiceTest.kt [hunk 4]
```diff
@@ -62,9 +81,32 @@ internal class RecallDocumentServiceTest {
...
+  @Test
+  fun `gets a document by recall ID and document ID`() {
+    val aDocumentId = UUID.randomUUID()
+    val aDocument = RecallDocument(
+      id = aDocumentId,
+      recallId = recallId.value,
+      category = RecallDocumentCategory.PART_A_RECALL_REPORT
+    )
+    val aRecallWithDocument = aRecall.copy(documents = setOf(aDocument))
+    val fileBytes = "Hello".toByteArray()
```
[the remark this answers is not in the document]
Comment (mr-ken) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/service/RecallDocumentServiceTest.kt:102]: > Please see few QQs left. Thanks.

Thanks for your review @AndyNewton. Your comments are noted concerning using randomly generated values but I can't see an established pattern to always use randomly generated values so I would argue to stick with the existing pattern.

**Comment by mr-ken** [comment]
> Please see few QQs left. Thanks.

Thanks for your review @AndyNewton. Your comments are noted concerning using randomly generated values but I can't see an established pattern to always use randomly generated values so I would argue to stick with the existing pattern.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/controller/RecallsController.kt [hunk 5]
```diff
@@ -78,6 +79,29 @@ class RecallsController(
...
+    try {
+      val (document, bytes) = recallDocumentService.getDocument(recallId, documentId)
+      return ResponseEntity.ok(
+        GetDocumentResponse(
+          documentId = documentId,
+          category = document.category,
+          content = Base64.getEncoder().encodeToString(bytes)
+        )
+      )
+    } catch (e: Throwable) {
```
[the remark this answers is not in the document]
Comment (mr-ken) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/controller/RecallsController.kt:97]: Noted - changing.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/service/RecallDocumentService.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (AndyNewton) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/service/RecallDocumentService.kt:33]: I'm suggesting simply that `id` in the first line of this extract be changed to `documentId` as per the 4th line:
```
data class AddDocumentResponse(val id: UUID)

data class GetDocumentResponse(
  val documentId: UUID,
  val category: RecallDocumentCategory,
  val content: String
)
```
It looks as if we did something exactly analogous last week in this segment:
```
data class RecallResponse(
  @Deprecated(message = "'id' deprecated in favour of 'recallId'", replaceWith = ReplaceWith("recallId"))
  val id: RecallId,
  val recallId: RecallId,
```

**Review by mikedunfordEE — APPROVED** [review]

**Pull request merged**